src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -338,67 +312,102 @@ public void onClose() {
         logger.debug("closed");
       }
       if (!isClosed) {
-        executorService.execute(new Runnable() {
-          @Override
-          public void run() {
-            onClosed();
-          }
-        });
+        // If the connection tear down was initiated by the higher-layer, isClosed will already
+        // be true. Nothing more to do in that case.
+        executorService.execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                onClosed();
+              }
+            });
       }
     }
 
     @Override
-    public void onError(final WebSocketException e) {
+    public void onError(final Throwable e) {
+      if (e.getCause() != null && e.getCause() instanceof EOFException) {
+        logger.error("WebSocket reached EOF", e);
+      } else {
+        logger.error("WebSocket error", e);
+      }
       executorService.execute(
           new Runnable() {
             @Override
             public void run() {
-              if (e.getCause() != null && e.getCause() instanceof EOFException) {
-                logger.debug("WebSocket reached EOF.");
-              } else {
-                logger.debug("WebSocket error.", e);
-              }
               onClosed();
```

Comment:
I don't think NettyWebsocketConnection always closes the connection before calling onError, so I wonder if this should be closeAndNotify() instead?

Reply:
That is what this method (`onClosed()`) does:

```
private void onClosed() {
    if (!isClosed) {
      if (logger.logsDebug()) {
        logger.debug("closing itself");
      }
      closeAndNotify();
    }
  }
```